### PR TITLE
Dag llapi

### PR DESCRIPTION
**Pull request by alonre24**
#456
Add the LLAPI for running a DAG.

### CODE REVIEW

File: src/DAG/dag_builder.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        RAI_SetError(err, RAI_EDAGBUILDER, "DAG op should start with: '|>' ");
+        return res;
+    }
+    RedisModuleString **argv = array_new(RedisModuleString *, 2);
+    while (token != NULL) {
+        RedisModuleString *RS_token = RedisModule_CreateString(NULL, token, strlen(token));
+        argv = array_append(argv, RS_token);
+        argc++;
+        token = strtok(NULL, " ");
+    }
```
Comment (DvirDukhan) [inline src/DAG/dag_builder.c:134]: move this to a dedicated function

File: src/DAG/dag_builder.c [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        const char *arg_string = RedisModule_StringPtrLen(argv[i], NULL);
+        if (strcmp(arg_string, "|>") == 0 && i < argc - 1) {
+            RAI_InitDagOp(&op);
+            rinfo->dagOps = array_append(rinfo->dagOps, op);
+            new_ops++;
+            op->argv = &argv[i + 1];
+        } else {
+            op->argc++;
+        }
+    }
```
Comment (DvirDukhan) [inline src/DAG/dag_builder.c:149]:
```
if (!strcasecmp(arg_string, "|>") && arg_pos < argc - 1) {
            RAI_DagOp *currentOp = _AddEmptyOp(rinfo);
            chainingOpCount++;
            int args_num = _CollectOpArgs(argv, argc, ++arg_pos, currentOp);
            arg_pos += args_num;
            continue;
        }
```
from dag_parser.c won't work here?

File: tests/module/LLAPI.c [hunk 3]
```diff
@@ -234,6 +235,69 @@ int RAI_llapi_scriptRun(RedisModuleCtx *ctx, RedisModuleString **argv, int argc)
...
+    run_info = RedisAI_DAGRunCtxCreate();
+    // Test the case of building the same DAG as in previous test, but when this time it should return with an error.
+    if(testSimpleDAGRun2Error(ctx, run_info) != LLAPIMODULE_OK) {
+        return RedisModule_ReplyWithSimpleString(ctx, "Simple DAG run2 error test failed");
+    }
+    run_info = RedisAI_DAGRunCtxCreate();
+    // Test the case of building DAG ops from string.
+    if(testBuildDAGFromString(ctx, run_info) != LLAPIMODULE_OK) {
+        return RedisModule_ReplyWithSimpleString(ctx, "Build DAG from string test failed");
+    }
```
Comment (DvirDukhan) [inline tests/module/LLAPI.c:282]: why the creation of run_info before every test? you can do it in the test itself

File: src/DAG/dag_execute.c [hunk 4]
```diff
@@ -285,16 +285,11 @@ RAI_Tensor *RAI_DAGOutputTensor(RAI_OnFinishCtx *finish_ctx, size_t index) {
...
-RAI_Error *RAI_DAGCopyOpStatus(RAI_OnFinishCtx *finish_ctx, size_t index) {
+RAI_Error *RAI_DAGGetError(RAI_OnFinishCtx *finish_ctx) {
     RedisAI_RunInfo *rinfo = (RedisAI_RunInfo *)finish_ctx;
-    RedisModule_Assert(index < rinfo->dagOpCount);
-    RAI_Error *err;
-    RAI_InitError(&err);
-    RAI_SetError(err, RAI_GetErrorCode(rinfo->dagOps[index]->err),
-                 RAI_GetError(rinfo->dagOps[index]->err));
-    return err;
+    return rinfo->err;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/DAG/dag_execute.c:294]: The rinfo obj. User should copy the error if he wants ownership, same apply for output tensors.

File: src/DAG/dag.c [hunk 5]
```diff
@@ -571,6 +571,9 @@ void RedisAI_DagRunSessionStep(RedisAI_RunInfo *rinfo, const char *devicestr) {
 
     if (currentOp->result != REDISMODULE_OK) {
         __atomic_store_n(rinfo->dagError, 1, __ATOMIC_RELAXED);
+        RAI_ContextWriteLock(rinfo);
+        RAI_SetError(rinfo->err, RAI_GetErrorCode(currentOp->err), RAI_GetError(currentOp->err));
+        RAI_ContextUnlock(rinfo);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/DAG/dag.c:576]: you right, added

File: src/redisai.c [hunk 6]
```diff
@@ -953,6 +953,7 @@ static int RedisAI_RegisterApi(RedisModuleCtx *ctx) {
     REGISTER_API(GetError, ctx);
     REGISTER_API(GetErrorOneLine, ctx);
     REGISTER_API(GetErrorCode, ctx);
+    REGISTER_API(SetError, ctx);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/redisai.c:956]: To allow copying the error details in the OnFinish CB

File: src/DAG/dag_parser.c [hunk 7]
```diff
@@ -283,19 +285,30 @@ int ParseDAGRunCommand(RedisAI_RunInfo *rinfo, RedisModuleCtx *ctx, RedisModuleS
...
+        for (size_t i = 0; i < array_len(dag_ops); i++) {
+            RAI_FreeDagOp(dag_ops[i]);
+        }
         goto cleanup;
     }
+    // After validating all the ops, insert them to the DAG.
+    for (size_t i = 0; i < array_len(dag_ops); i++) {
+        rinfo->dagOps = array_append(rinfo->dagOps, dag_ops[i]);
+    }
+    rinfo->dagOpCount = array_len(rinfo->dagOps);
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/DAG/dag_parser.c:304]: yes I will add this logic to the ParseDAGOps function

File: tests/module/LLAPI.c (hunk 3 above)
Comment (alonre24) [reply tests/module/LLAPI.c:282]: done

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**